File: library/src/main/java/timber/log/ConsoleTree.java
package timber.log;

/**
 * Created by flisar on 20.10.2016.
 */

import android.util.Log;

import com.michaelflisar.lumberjack.L;
import com.michaelflisar.lumberjack.LogUtil;

public class ConsoleTree extends BaseTree
{
    public ConsoleTree(boolean combineTags, boolean withLink)
    {
        super(combineTags, withLink);
    }

    @Override
    protected void log(int priority, String tag, String message, Throwable t) {

        if (mStackData != null && mWithLink)
            message = mStackData.appendLink(message);

        if (message.length() < MAX_LOG_LENGTH) {
            logLine(priority, tag, message);
            return;
        }

        // Split by line, then ensure each line can fit into Log's maximum length.
        for (int i = 0, length = message.length(); i < length; i++) {
            int newline = message.indexOf('\n', i);
            newline = newline != -1 ? newline : length;
            do {
                int end = Math.min(newline, i + MAX_LOG_LENGTH);
                String part = message.substring(i, end);
                L.getFormatter().formatLine(tag, part);
                i = end;
            } while (i < newline);
        }
    }

    private void logLine(int priority, String tag, String message)
    {
        String logMessage = L.getFormatter().formatLine(tag, message);

        if (priority == Log.ASSERT) {
            Log.wtf(tag, logMessage);
        } else {
            Log.println(priority, tag, logMessage);
        }
    }
}

File: library/src/main/java/timber/log/ConsoleTree.kt
package timber.log

import android.annotation.SuppressLint
import android.util.Log
import com.michaelflisar.lumberjack.data.StackData
import kotlin.math.min

/**
 * Created by flisar on 17.01.2019.
 */

class ConsoleTree(
    private val appendClickableLink: Boolean = true,
    private val replaceBracketsIfAppendableLinks: Boolean = false
) : BaseTree() {

    companion object {
        private const val MAX_LOG_LENGTH = 4000
    }

    override fun log(
        priority: Int,
        prefix: String,
        message: String,
        t: Throwable?,
        stackData: StackData
    ) {
        val msg = if (replaceBracketsIfAppendableLinks && appendClickableLink) {
            message
                .replace("(", "{")
                .replace(")", "}")
        } else message
        val fullMessage = if (appendClickableLink) appendLink(msg, stackData) else msg

        if (fullMessage.length < MAX_LOG_LENGTH) {
            logLine(priority, prefix, fullMessage)
            return
        }

        // Split by line, then ensure each line can fit into Log's maximum length.
        var i = 0
        val length = fullMessage.length
        while (i < length) {
            var newline = fullMessage.indexOf('\n', i)
            newline = if (newline != -1) newline else length
            do {
                val end = min(newline, i + MAX_LOG_LENGTH)
                val part = fullMessage.substring(i, end)
                logLine(priority, prefix, part)
                i = end
            } while (i < newline)
            i++
        }
    }

    @SuppressLint("LogNotTimber")
    private fun logLine(priority: Int, prefix: String, message: String) {
        val logMessage = formatLine(prefix, message)
        if (priority == Log.ASSERT) {
            Log.wtf(prefix, logMessage)
        } else {
            Log.println(priority, prefix, logMessage)
        }
    }

    private fun appendLink(message: String, stackData: StackData): String {
        val link = stackData.getLink()
        return "$message $link"
    }
}